### PR TITLE
Update ContentParser API

### PR DESCRIPTION
**Pull request by Ethan-Arrowood**
Ref #525 
Need to update the ContentParser `done()` method. Originally it only took one argument `body`. Split into `error, data`, then promisify the method.
- [x] Split body parameter into error, data in done() method
- [x] Promisify API
- [x] Add test for async case
- [x] Update docs
- [x] Update docs again with async changes

### CODE REVIEW

File: lib/ContentTypeParser.js [hunk 1]
```diff
@@ -48,10 +48,10 @@ ContentTypeParser.prototype.getHandler = function (contentType) {
 ContentTypeParser.prototype.run = function (contentType, handler, context, params, req, res, query) {
   this.getHandler(contentType)(req, done)
 
-  function done (body) {
-    if (body instanceof Error) {
+  function done (error, body) {
+    if (error instanceof Error) {
```
Comment (delvedor) [inline lib/ContentTypeParser.js]: This check is not needed,
```js
if (error)
```
should be enough :)

File: test/custom-parser.test.js [hunk 2]
```diff
@@ -136,7 +136,7 @@ test('contentTypeParser should handle errors', t => {
   })
 
   fastify.addContentTypeParser('application/jsoff', function (req, done) {
-    done(new Error('kaboom!'))
+    done(new Error('kaboom!'), {})
```
Comment (allevo) [inline test/custom-parser.test.js]: Could you remove the `{}`?

**Comment by Ethan-Arrowood** [comment]
Will do!

File: test/custom-parser.test.js [hunk 3]
```diff
@@ -193,8 +193,8 @@ test('contentTypeParser should support encapsulation, second try', t => {
 
     instance.addContentTypeParser('application/jsoff', function (req, done) {
       jsonParser(req, function (err, body) {
-        if (err) return done(err)
-        done(body)
+        // if (err) return done(err)
```
Comment (mcollina) [inline test/custom-parser.test.js]: can you remove the comment?

File: test/custom-parser.test.js [hunk 4]
```diff
@@ -230,8 +230,8 @@ test('contentTypeParser shouldn\'t support request with undefined "Content-Type"
 
   fastify.addContentTypeParser('application/jsoff', function (req, done) {
     jsonParser(req, function (err, body) {
-      if (err) return done(err)
-      done(body)
+      // if (err) return done(err)
```
Comment (mcollina) [inline test/custom-parser.test.js]: can you remove the comment?

File: test/custom-parser.test.js [hunk 5]
```diff
@@ -349,14 +349,14 @@ test('catch all content type parser should not interfere with other conte type p
...
-      done(data)
+      done(null, data)
     })
   })
 
   fastify.addContentTypeParser('application/jsoff', function (req, done) {
     jsonParser(req, function (err, body) {
-      if (err) return done(err)
-      done(body)
+      // if (err) return done(err)
```
Comment (mcollina) [inline test/custom-parser.test.js]: can you remove the comment?

File: test/custom-parser.test.js [hunk 6]
```diff
@@ -87,8 +87,8 @@ test('contentTypeParser should handle multiple custom parsers', t => {
 
   function customParser (req, done) {
     jsonParser(req, function (err, body) {
-      if (err) return done(err)
-      done(body)
+      // if (err) return done(err)
```
Comment (mcollina) [inline test/custom-parser.test.js]: can you remove the comment?

File: test/custom-parser.test.js (hunk 5 above)
Comment (Ethan-Arrowood) [reply test/custom-parser.test.js]: oops forgot about all these!

**Comment by Ethan-Arrowood** [comment]
Can I have some additional pointers on integrating async/await?

**Comment by delvedor** [comment]
@Ethan-Arrowood checkout what we did in handleRequest.js :)

https://github.com/fastify/fastify/blob/ed65c02c9fa5084603646c99405a65e55763f40a/lib/handleRequest.js#L123-L136

**Comment by Ethan-Arrowood** [comment]
Still a little lost here. Should I be implementing this in: https://github.com/fastify/fastify/blob/ed65c02c9fa5084603646c99405a65e55763f40a/lib/ContentTypeParser.js#L48-L58

Also is the following valid? If not what should the async version of this api look like?
```javascript
fastify.addContentTypeParser('application/jsoff', async function (req, done) {
  await jsonParser(req).then(function (err, body) {
    done(err, body)
  })
})
```

**Comment by mcollina** [comment]
I think the API should be:

```js
fastify.addContentTypeParser('application/jsoff', async function (req) {
  return await jsonParser(req)
})
```

**Comment by Ethan-Arrowood** [comment]
The linter throws a `Redundant use of `await` on a return value` when I do that. Should I suppress this error?

**Comment by Ethan-Arrowood** [comment]
And so as I understand this, given `await jsonParser(req)` the `jsonParser(req)` part must be a promise that either rejects or resolves. Because the literal method `jsonParser` is from `fast-json-body`, is the goal to wrap it in a promise? And I'm assuming I have to do this in the `ContentTypeParser.js` file, if so where?

**Comment by mcollina** [comment]
I mean something completely different here, absolutely not related to json:

```js
fastify.addContentTypeParser('application/something-else', async function (req) {
  const result = await doSomethingAsync(req)
  return result + 'whatever'
})
```

**Comment by delvedor** [comment]
What you should do is something like the following:
```js
 ContentTypeParser.prototype.run = function (contentType, handler, context, params, req, res, query) { 
   var result = this.getHandler(contentType)(req, done)
   if (result && typeof result === 'function') {
     result.then(body => done(null, body)).catch(err => done(err, null))
   } 
  
   function done (err, body) { 
     if (err) { 
       const reply = new context.Reply(res, context, null) 
       return reply.send(err) 
     } 
     handler(context, params, req, res, body, query) 
   } 
 } 
```

**Comment by Ethan-Arrowood** [comment]
It was the `var result = this.getHandler(...)` line that I was missing. I couldn't figure out where was the appropriate place to add the `.then().catch()` statements. Wonderful! I'll add this and try my hand at including a test case!

**Comment by Ethan-Arrowood** [comment]
Okay I've pushed the api updates. Testing is a headache but I'm determined to figure it out. I'm going to see if I can find any examples in the other test files of testing async operations

**Comment by delvedor** [comment]
To test the promises part you can try to [`promisify`](https://www.npmjs.com/package/util-promisify) the json body parser.
Remember to test the *async-await* code only in Node ≥ v8.
https://github.com/fastify/fastify/blob/ed65c02c9fa5084603646c99405a65e55763f40a/test/async-await.test.js#L6-L11

**Comment by Ethan-Arrowood** [comment]
The builds are failing because of node version even though I have the `semver.lt(process.versions.node, '8.0.0')` not sure what to do.

File: lib/ContentTypeParser.js [hunk 7]
```diff
@@ -46,14 +46,16 @@ ContentTypeParser.prototype.getHandler = function (contentType) {
 }
 
 ContentTypeParser.prototype.run = function (contentType, handler, context, params, req, res, query) {
-  this.getHandler(contentType)(req, done)
-
-  function done (body) {
-    if (body instanceof Error) {
+  var result = this.getHandler(contentType)(req, done)
+  if (result && typeof result === 'function') {
```
Comment (mcollina) [inline lib/ContentTypeParser.js]: I would check `result.then`.

File: lib/ContentTypeParser.js [hunk 8]
```diff
@@ -46,14 +46,16 @@ ContentTypeParser.prototype.getHandler = function (contentType) {
...
+    result.then(body => done(null, body)).catch(err => done(err, null))
+  }
+  function done (error, data) {
```
Comment (delvedor) [inline lib/ContentTypeParser.js]: Can you rename `data` in `body`?

File: lib/ContentTypeParser.js (hunk 7 above)
Comment (delvedor) [reply lib/ContentTypeParser.js]: My fault, was I typo in my suggestion above :D

File: docs/ContentTypeParser.md [hunk 9]
```diff
@@ -9,9 +9,14 @@ As with the other APIs, `addContentTypeParser` is encapsulated in the scope in w
 ```js
 fastify.addContentTypeParser('application/jsoff', function (req, done) {
   jsoffParser(req, function (err, body) {
-    done(err || body)
+    done(err, body)
   })
 })
+// async also supported in Node versions > 8.0.0
+fastify.addContentTypeParser('application/jsoff', async function (req, done) {
```
Comment (mcollina) [inline docs/ContentTypeParser.md]: you should remove `done` from the function definition.

File: test/custom-parser.test.js (hunk 2 above)
Comment (mcollina) [reply test/custom-parser.test.js]: @Ethan-Arrowood ^

File: docs/ContentTypeParser.md [hunk 10]
```diff
@@ -9,9 +9,14 @@ As with the other APIs, `addContentTypeParser` is encapsulated in the scope in w
 ```js
 fastify.addContentTypeParser('application/jsoff', function (req, done) {
   jsoffParser(req, function (err, body) {
-    done(err || body)
+    done(err, body)
   })
 })
+// async also supported in Node versions > 8.0.0
```
Comment (delvedor) [inline docs/ContentTypeParser.md]: async also supported in Node versions **>=** 8.0.0

**Comment by mcollina** [comment]
You can squash when landing!
Il giorno mar 12 dic 2017 alle 18:26 Tomas Della Vedova <
notifications@github.com> ha scritto:

**Comment by Ethan-Arrowood** [comment]
Yeah there should be a 'Squash and Merge` option. If not I can give it a shot via the cmnd line.

**Comment by Ethan-Arrowood** [comment]
@delvedor all squashed and ready to go 👍 
Thanks for letting me work through this one guys! Learned a lot of awesome stuff 😄

**Pull request merged**